From 028122c6cef73c3e35c378136f44174129a62d82 Mon Sep 17 00:00:00 2001
From: =?UTF-8?q?R=C3=A9mi=20Alvergnat?= <toilal.dev@gmail.com>
Date: Fri, 27 Sep 2019 01:24:32 +0200
Subject: [PATCH 1/2] Add support for more column names

This adds support for column names starting with numbers, containing dashes and spaces.
---
 src/Adapter/Platform/AbstractPlatform.php |  2 +-
 src/Sql/AbstractSql.php                   | 20 +++++++++++++++++---
 2 files changed, 18 insertions(+), 4 deletions(-)

diff --git a/src/Adapter/Platform/AbstractPlatform.php b/src/Adapter/Platform/AbstractPlatform.php
index 42b885751..4d874ba30 100644
--- a/src/Adapter/Platform/AbstractPlatform.php
+++ b/src/Adapter/Platform/AbstractPlatform.php
@@ -29,7 +29,7 @@ abstract class AbstractPlatform implements PlatformInterface
     /**
      * @var string
      */
-    protected $quoteIdentifierFragmentPattern = '/([^0-9,a-z,A-Z$_:])/i';
+    protected $quoteIdentifierFragmentPattern = '/([^0-9,a-z,A-Z$_\-:])/i';
 
     /**
      * {@inheritDoc}
diff --git a/src/Sql/AbstractSql.php b/src/Sql/AbstractSql.php
index f8935ca44..71c61def3 100644
--- a/src/Sql/AbstractSql.php
+++ b/src/Sql/AbstractSql.php
@@ -428,9 +428,23 @@ abstract class AbstractSql implements SqlInterface
         if ($column === null) {
             return 'NULL';
         }
-        return $isIdentifier
-                ? $fromTable . $platform->quoteIdentifierInFragment($column)
-                : $platform->quoteValue($column);
+
+        if ($isIdentifier) {
+            $startsWithNumber = preg_match('/^[0-9]+/', $column);
+            $hasSpaceOrDash = preg_match('/^(.+)(\s|-)+(.+)$/i', $column);
+            $isUnicode = strlen($column) != mb_strlen($column);
+            if ($startsWithNumber || $hasSpaceOrDash || $isUnicode) {
+                $column = $platform->quoteIdentifier($column);
+            } else {
+                $column = $platform->quoteIdentifierInFragment($column);
+            }
+
+            $column = $fromTable . $column;
+        } else {
+            $column = $platform->quoteValue($column);
+        }
+
+        return $column;
     }
 
     /**
-- 
2.17.1

